### PR TITLE
Adding HTTPS on Prod version

### PR DESCRIPTION
**Pull request by RichtXO**
**Features Added**

I added certs and open both port 80 and 443. Port 80 will redirect to port 443. Also made it auto-generate the OpenSSL certificate every time it reruns. 

**Additional Info**

https://tech.osteel.me/posts/docker-for-local-web-development-part-5-https-all-the-things#generating-the-certificate
https://www.digitalocean.com/community/tutorials/how-to-create-a-self-signed-ssl-certificate-for-nginx-in-ubuntu-16-04

### CODE REVIEW

**Comment by 123joshuawu** [comment]
The autogenerating openssl is cool

Seems like the redirect is redirecting to `localhost:443` which isn't working on the pr-deploy

Also, just wondering, how would we specify custom ssl certificates?

**Comment by RichtXO** [comment]
Oh, that's weird... I'll double-check with my end on this. 
Just wondering, did you open 443 for the thing that generates the deploy link?

**Comment by 123joshuawu** [comment]
Ah I did not, only the `:80` to `:443` redirect

**Comment by RichtXO** [comment]
But still... If that was the case, we shouldn't even hit get into port 80. We should just get an error message or something...

**Comment by RichtXO** [comment]
Wait, what does that mean?

**Comment by 123joshuawu** [comment]
So the issue is the `pr-deploy` is not overriding the `$HOST` env var for the `web` container so it defaults to `localhost`. Thus when you go to port 80, it returns a 301 redirect to `https://localhost:443` which fails

**Comment by RichtXO** [comment]
So now said that's a deployment issue? Nothing wrong with the code. It's just that this deployment thing isn't (like you said) not overriding.

**Comment by 123joshuawu** [comment]
Yes, unfortunately we'll need to fix that before we merge or its going to break on subsequent pull requests. tagging @jshom re: the `pr-cd` workflow

**Comment by jshom** [comment]
1. PR-CD can be fixed by going around redirects through providing link in github as HTTPS instead of HTTP in https://github.com/YACS-RCOS/yacs.n/blob/master/ops/provision.js#L73
2. Link certificates through docker volume for prod deploy and other environments

**Comment by jshom** [comment]
oi, the build info link... this is fine for now, will make another issue for this, no need to fix it in this PR (this would need to stick to http)

**Comment by RichtXO** [comment]
Yeah. But this is weird. I think 3 days ago, I couldn't get into port 443 on the old URL: `173.255.231.178/`. I typed the URL with the HTTPS and didn't go through. It may have been my cookies.

**Comment by jshom** [comment]
> Yeah. But this is weird. I think 3 days ago, I couldn't get into port 443 on the old URL: `45.33.71.174`. I typed the URL with the HTTPS and didn't go through. It may have been my cookies.

Ye maybe, I think that could have been a flaky build too.

**Comment by jshom** [comment]
^ https://45.79.183.91/?semester=FALL%202020 use direct https link

File: src/web/Dockerfile [hunk 1]
```diff
@@ -12,6 +12,7 @@ RUN mkdir /app
 COPY --from=build-stage /app/dist /app
 COPY nginx.conf /etc/nginx/nginx.template.conf
 COPY scripts/entrypoint.sh /usr/local/bin/entrypoint.sh
+ADD certificate/ /etc/nginx/certificate
```
Comment (123joshuawu) [inline src/web/Dockerfile:15]: I believe its preferable to use `COPY` over `ADD`?

File: src/web/scripts/entrypoint.sh [hunk 2]
```diff
@@ -13,6 +13,18 @@ envsubst '\$HOST' < \
   /etc/nginx/nginx.template.conf > \
   /etc/nginx/nginx.conf
 
+
+# If SSL Certificate folder isn't present, generate one
+if [ ! -d "/etc/nginx/certificate" ];then
```
Comment (123joshuawu) [inline src/web/scripts/entrypoint.sh:18]: Can also check if there are certificates in the folder? What do you think

**Review by 123joshuawu — COMMENTED** [review]
Very cool, nice solution!

File: src/web/scripts/entrypoint.sh (hunk 2 above)
Comment (RichtXO) [reply src/web/scripts/entrypoint.sh:18]: Oh, didn't think of that! Easy fix!

**Comment by jshom** [comment]
Small thing from a glance, seeing file `src/web/certificate/READMD.md` changed :) did u mean `src/web/certificate/README.md`. Gonna test with custom cert in a bit, sorry bout the delay on the review here. Looks great!

**Comment by RichtXO** [comment]
> Small thing from a glance, seeing file `src/web/certificate/READMD.md` changed :)

What changed on the README? I was planning to add more stuff on how to obtain a cert and issue it to a CA for free.

>  did u mean `src/web/certificate/README.md`. Gonna test with custom cert in a bit, sorry bout the delay on the review here.

It's alright! I didn't know if what I implemented was what you guys had in mind!

**Comment by jshom** [comment]
https://www.ssllabs.com/ssltest/analyze.html?d=li1277%2d238.members.linode.com&s=45.79.178.238&latest

**Comment by jshom** [comment]
WOOOHOOO, thing works. Gonna open new issue to address nginx tweaks for protocols & ciphers to address the test.

File: src/web/certificate/READMD.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# SSL Certificate for yacs-web
+
+The public/private keypairs are default and should be changed before deployment to production!
+Generate your own certificate and verify it with a CA!
```
Comment (jshom) [inline src/web/certificate/READMD.md:4]: Plz rename file to README.md :) not READMD.md

File: src/web/nginx.conf [hunk 4]
```diff
@@ -13,10 +13,24 @@ http {
   default_type application/octet-stream;
   keepalive_timeout  65;
 
-  server {
+
```
Comment (jshom) [inline src/web/nginx.conf:16]: remove this newline

File: src/web/scripts/entrypoint.sh [hunk 5]
```diff
@@ -13,6 +13,20 @@ envsubst '\$HOST' < \
   /etc/nginx/nginx.template.conf > \
   /etc/nginx/nginx.conf
 
+
+# If SSL Certificate folder isn't present, generate one
+if [ ! -f /etc/nginx/certificate/$HOST.crt ] &&
+[ ! -f /etc/nginx/certificate/$HOST.csr ] &&
```
Comment (jshom) [inline src/web/scripts/entrypoint.sh:19]: Remove this line, not necessary for providing in Nginx

File: src/web/nginx.conf [hunk 6]
```diff
@@ -13,10 +13,24 @@ http {
...
+  server {
+    listen 443 ssl http2;
+    listen [::]:443 ssl http2;
+    
+    ssl_certificate      /etc/nginx/certificate/${HOST}.crt;
+    ssl_certificate_key  /etc/nginx/certificate/${HOST}.key;
+
     server_name ${HOST};
 
+
```
Comment (jshom) [inline src/web/nginx.conf:33]: Remove newline

File: ops/provision.js [hunk 7]
```diff
@@ -69,7 +69,8 @@ const infraSync = async () => {
   // print ip
   // USED IN GITHUB ACTIONS PIPELINE TO SHOW MESSAGE IN PR
   // DO NOT REFORMAT
-  console.log(`http://${info.ipv4[0]}`)
+  console.log(`https://${info.ipv4[0]}`)
+
```
Comment (jshom) [inline ops/provision.js:73]: Remove newline

**Review by jshom — APPROVED** [review]
After removal of the csr check line for provided key/cert, and small whitespace fixes; good for merge!

**Review by 123joshuawu — APPROVED** [review]

**Pull request merged**